### PR TITLE
Schema Validation for Charge and Charge Link (post merge)

### PR DESCRIPTION
**Pull request by defectiveAi**
Because of project structure changes, I reapplied the code from previous PR in a new branch.

### CODE REVIEW

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandConverter.cs [hunk 1]
```diff
@@ -242,7 +240,11 @@ private async Task<Instant> ParseTimeIntervalAsync(XmlReader reader, Instant sta
             {
                 if (reader.Is(CimChargeCommandConstants.TimeIntervalStart, CimChargeCommandConstants.Namespace))
                 {
-                    return Instant.FromDateTimeUtc(reader.ReadElementContentAsDateTime());
+                    var cimTimeInterval = await reader
+                        .ReadElementContentAsStringAsync()
+                        .ConfigureAwait(false);
+
+                    return Instant.FromDateTimeOffset(DateTimeOffset.Parse(cimTimeInterval));
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandConverter.cs:247]: Out of curiosity: Why is this change necessary?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandDeserializer.cs [hunk 2]
```diff
@@ -12,33 +12,34 @@
...
+using Energinet.DataHub.Core.Schemas;
+using Energinet.DataHub.Core.SchemaValidation;
+using Energinet.DataHub.Core.SchemaValidation.Extensions;
 using GreenEnergyHub.Charges.Domain.Dtos.ChargeCommands;
-using GreenEnergyHub.Charges.Infrastructure.Core.MessagingExtensions.Serialization;
 
 namespace GreenEnergyHub.Charges.Infrastructure.CimDeserialization.ChargeBundle
 {
-    public class ChargeCommandDeserializer : MessageDeserializer<ChargeCommandBundle>
+    public sealed class ChargeCommandDeserializer : SchemaValidatingMessageDeserializer<ChargeCommandBundle>
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandDeserializer.cs:24]: Out of curiosity: Is there a particular reason for making this class sealed? Do you think we should seal all our classes (if yes, why)?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestFiles/Charges/InvalidTariffDocument.xml [hunk 3]
```diff
@@ -41,8 +41,8 @@ limitations under the License.
                 <cim:Series_Period>
                     <cim:resolution>P1D</cim:resolution>
                     <cim:timeInterval>
-                        <cim:start>2020-09-26T22:00:00Z</cim:start>
-                        <cim:end>2020-09-26T22:00:00Z</cim:end>
+                        <cim:start>2020-09-26T22:00Z</cim:start>
+                        <cim:end>2020-09-26T22:00Z</cim:end>
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestFiles/Charges/InvalidTariffDocument.xml:45]: Is this the required date time format in CIM?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestHelpers/EmbeddedResourceHelper.cs [hunk 4]
```diff
@@ -35,7 +36,11 @@ private static string ReplaceMergeFields(IClock clock, string file)
         {
             var currentInstant = clock.GetCurrentInstant();
             var now = currentInstant.ToString();
-            var inThirtyoneDays = currentInstant.Plus(Duration.FromDays(31)).ToString();
+            var inThirtyoneDays = currentInstant.Plus(Duration.FromDays(31));
+
+            // cim:timeInterval does not allow seconds.
+            var ymdhmTimeInterval = inThirtyoneDays
+                .ToString("yyyy-MM-dd\\THH:mm\\Z", CultureInfo.InvariantCulture);
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestHelpers/EmbeddedResourceHelper.cs:43]: Seems like a candidate to be put somewhere more easily discovered and reused as we would like to only maintain this specific CIM formatting in a single place.

**Review by BjarkeMeier — COMMENTED** [review]
LGTM, only minor comments and questions.
However, should we have tests of each schema validation? Why or why not?

**Comment by defectiveAi** [comment]
> LGTM, only minor comments and questions. However, should we have tests of each schema validation? Why or why not?

I have only added two basic UTs that check if "Bad Request" is returned if schema validation fails. So there are now 4 tests in total - one for positive outcome and one for negative outcome for each of the two schemas.

I would probably not test the individual implementations of the schemas, but instead have a UT-covered validation step somewhere in application layer. That way you will only have one set of tests across all formats. The counter-argument is that the schemas _is_ the validation step, but there will be multiple sources of truth (one for xml and one for json) and since the schemas are provided for you, the UTs will lock down an implementation you have no control over and may change a lot.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs [hunk 5]
```diff
@@ -57,21 +67,35 @@ public class ChargeIngestion
...
+        private static async Task<HttpResponseData> CreateErrorResponseAsync(
+            HttpRequestData req,
+            ErrorResponse errorResponse)
         {
-            var response = req.CreateResponse();
-            await response.WriteAsJsonAsync(messageResult);
+            var response = req.CreateResponse(HttpStatusCode.BadRequest);
+            await errorResponse.WriteAsXmlAsync(response.Body).ConfigureAwait(false);
+            return response;
+        }
```
[the remark this answers is not in the document]
Comment (defectiveAi) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs]: I have added IHttpFunctionResponseBuilder. The implementation is mostly empty at the moment (but avoids repetition of code), but could grow over time; e.g. support for both xml and json. Feedback is welcome.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandDeserializer.cs (hunk 2 above)
Comment (defectiveAi) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandDeserializer.cs:24]: I usually seal all my classes by default. Writing classes with extensibility in mind is hard, and the use-case is often not there initially, and sealing by default forces the next person to explicitly unsealing it (and hopefully consider the consequences of doing so). It also promotes composition over inheritance by default.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestFiles/Charges/InvalidTariffDocument.xml (hunk 3 above)
Comment (defectiveAi) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestFiles/Charges/InvalidTariffDocument.xml:45]: Yes, and it was so odd that I asked Jan about it. The answer is that timeseries and priceseries (only those two) are going to use this format without seconds.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestHelpers/EmbeddedResourceHelper.cs (hunk 4 above)
Comment (defectiveAi) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestHelpers/EmbeddedResourceHelper.cs]: I moved the format string into a constant variable, so that the name could help indicate its purpose. I am unsure of where else to put it, as I do not think there should be a use-case for this format outside of generating test files?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestFiles/Charges/InvalidTariffDocument.xml (hunk 3 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestFiles/Charges/InvalidTariffDocument.xml:45]: Interesting...

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandDeserializer.cs (hunk 2 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure/CimDeserialization/ChargeBundle/ChargeCommandDeserializer.cs:24]: I like - and learn. Thank you :-)

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestHelpers/EmbeddedResourceHelper.cs (hunk 4 above)
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.IntegrationTests/TestHelpers/EmbeddedResourceHelper.cs]: Sounds reasonable.

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Function/IHttpFunctionResponseBuilder.cs [hunk 6]
```diff
@@ -0,0 +1,31 @@
...
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+using System.Threading.Tasks;
+using Energinet.DataHub.Core.SchemaValidation.Errors;
+using Microsoft.Azure.Functions.Worker.Http;
+
+namespace GreenEnergyHub.Charges.Infrastructure.Core.Function
+{
+    public interface IHttpFunctionResponseBuilder
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Function/IHttpFunctionResponseBuilder.cs:21]:
```suggestion
    public interface IHttpResponseBuilder
```

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Function/IHttpFunctionResponseBuilder.cs [hunk 7]
```diff
@@ -0,0 +1,31 @@
...
+    {
+        Task<HttpResponseData> CreateAcceptedResponseAsync<T>(
```
Comment (BjarkeMeier) [inline source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Function/IHttpFunctionResponseBuilder.cs:23]: Perhaps document the HTTP status code of the accepted (I anticipate 202) and error (I anticipate 404).
Btw, should the other method be named `CreateBadRequestResponseAsync`?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs (hunk 5 above)
[the remark this answers is not in the document]
Comment (BjarkeMeier) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.FunctionHost/Charges/ChargeIngestion.cs]: Nice, but see new comments on the implementation.

**Comment by BjarkeMeier** [comment]
> > LGTM, only minor comments and questions. However, should we have tests of each schema validation? Why or why not?
> 
> I have only added two basic UTs that check if "Bad Request" is returned if schema validation fails. So there are now 4 tests in total - one for positive outcome and one for negative outcome for each of the two schemas.
> 
> I would probably not test the individual implementations of the schemas, but instead have a UT-covered validation step somewhere in application layer. That way you will only have one set of tests across all formats. The counter-argument is that the schemas _is_ the validation step, but there will be multiple sources of truth (one for xml and one for json) and since the schemas are provided for you, the UTs will lock down an implementation you have no control over and may change a lot.

Maybe I just don't understand, but the essence of my question is: Is the charges domain responsible for each schema validation rule or is it the responsibility delegated to the schema validation package?

File: source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Function/IHttpFunctionResponseBuilder.cs (hunk 7 above)
Comment (defectiveAi) [reply source/GreenEnergyHub.Charges/source/GreenEnergyHub.Charges.Infrastructure.Core/Function/IHttpFunctionResponseBuilder.cs]: Good call, it should have been 202 Accepted.
For validation errors, the result is 400 Bad Request (I have renamed the method).
Also, I tried, but it is not feasible to mock the HttpResponseData, as it uses extension methods and FunctionContext that are not mockable.

**Comment by defectiveAi** [comment]
> > > LGTM, only minor comments and questions. However, should we have tests of each schema validation? Why or why not?
> > 
> > 
> > I have only added two basic UTs that check if "Bad Request" is returned if schema validation fails. So there are now 4 tests in total - one for positive outcome and one for negative outcome for each of the two schemas.
> > I would probably not test the individual implementations of the schemas, but instead have a UT-covered validation step somewhere in application layer. That way you will only have one set of tests across all formats. The counter-argument is that the schemas _is_ the validation step, but there will be multiple sources of truth (one for xml and one for json) and since the schemas are provided for you, the UTs will lock down an implementation you have no control over and may change a lot.
> 
> Maybe I just don't understand, but the essence of my question is: Is the charges domain responsible for each schema validation rule or is it the responsibility delegated to the schema validation package?

If you mean, whether the SchemaValidatingReader is covered by UTs, then yes. The package ensures that it works. If you ask whether the XSDs are covered by UTs, then no and there are currently no plans to do that.

**Review by BjarkeMeier — APPROVED** [review]
LGTM
